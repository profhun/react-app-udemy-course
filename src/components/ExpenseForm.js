import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();
now.format('YYYY-MM-DD');

export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            focused: false,
            error: undefined
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({
            description
        }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (amount.match(/^(\d+(\.\d{0,2})?)?$/)){
            this.setState(() => ({
                amount
            }));
        }
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }));
    };

    onDateChange = (createdAt) => {
        if (createdAt){
            this.setState(() => ({
                createdAt
            }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            focused
        }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount){
            this.setState(() => ({
                error: 'Bíííííztos?'
            }));
        } else {
            this.setState(() => ({
                error: undefined
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };

    render(props){
        return (
            <div>
                {this.state.error && <div>{this.state.error}</div>}
                <form
                    onSubmit={this.onSubmit}
                >
                    <input
                        type="text"
                        name="description"
                        placeholder="description"
                        value={this.state.description}
                        autoFocus
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        name="amount"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.focused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                    />
                    <textarea
                        name="note"
                        placeholder="note"
                        onChange={this.onNoteChange}
                    >
                        {this.state.note}
                    </textarea>
                    <button type="submit">Send</button>
                </form>
            </div>
        )
    }
}
