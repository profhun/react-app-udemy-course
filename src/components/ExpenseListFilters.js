import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends React.Component{
    state = {
        focusedInput: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (focusedInput) => {
        this.setState(() => ({
            focusedInput
        }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        if(e.target.value === "date"){
            this.props.sortByDate();
        }
        else if(e.target.value === "amount"){
            this.props.sortByAmount();
        }
    };

    render(){
        return (
            <div>
                <input
                    type="text"
                    placeholder="this is the best place for search"
                    defaultValue={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focusedInput}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch, props) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);