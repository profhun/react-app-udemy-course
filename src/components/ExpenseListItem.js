import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <li>
        <h4>{description}</h4>
        <div>{amount}</div>
        <div>{createdAt}</div>
        <Link to={`/edit/${id}`}>edit</Link>
    </li>
);

export default ExpenseListItem;