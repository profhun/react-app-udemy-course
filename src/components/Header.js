import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = (props) => (
    <header>
        <h1>Kiad√°sok</h1>
        <div>
            <NavLink to="/" activeClassName="is-active" exact={true}>~</NavLink>
            <NavLink to="/add" activeClassName="is-active">+</NavLink>
            <NavLink to="/help" activeClassName="is-active">?</NavLink>
        </div>
    </header>
);

export default Header;