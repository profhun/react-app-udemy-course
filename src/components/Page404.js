import React from 'react';
import { Link } from 'react-router-dom';

const Page404 = (props) => (
    <div>
        <h1>404</h1>
        <Link to="/">Back to home</Link>
    </div>
);

export default Page404;