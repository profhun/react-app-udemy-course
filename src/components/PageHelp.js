import React from 'react';

const PageHelp = (props) => (
    <div>help</div>
);

export default PageHelp;