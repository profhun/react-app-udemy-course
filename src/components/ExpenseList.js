import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        <h2>Expense List</h2>
        <p>Tételek száma: {props.expenses.length}</p>
        {
            props.expenses.length === 0 ? (
                <h4>Nincs semmi kiadásod. Örüljél!</h4>
            ) : (
                <ol>
                    {props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense}/>
                    })}
                </ol>
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);
