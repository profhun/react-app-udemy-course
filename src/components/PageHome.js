import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'

const PageHome = (props) => (
    <div>
        <h1>
            Hoooome
        </h1>
        <ExpenseListFilters />
        <ExpenseList {...props} />
    </div>

);

export default PageHome;