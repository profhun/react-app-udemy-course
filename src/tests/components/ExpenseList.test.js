import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses'

test('render/component ExpenseList', () => {

    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper.find('h2').text()).toBe("Expense List");
    expect(wrapper).toMatchSnapshot();
});


test('render/component ExpenseList with no item', () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper.find('h4').text()).toBe("Nincs semmi kiadásod. Örüljél!");
    expect(wrapper).toMatchSnapshot();
});
