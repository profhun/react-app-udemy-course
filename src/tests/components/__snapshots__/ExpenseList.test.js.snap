// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render/component ExpenseList 1`] = `
<div>
  <h2>
    Expense List
  </h2>
  <p>
    Tételek száma: 
    4
  </p>
  <ol>
    <ExpenseListItem
      amount={400}
      createdAt={-864000000}
      description="zokni a lábon"
      id="zokni"
      key="zokni"
      note=""
    />
    <ExpenseListItem
      amount={300}
      createdAt={-777600000}
      description="hawaii pizza"
      id="pizza"
      key="pizza"
      note=""
    />
    <ExpenseListItem
      amount={200}
      createdAt={604800000}
      description="BBQ"
      id="hamburger"
      key="hamburger"
      note=""
    />
    <ExpenseListItem
      amount={100}
      createdAt={691200000}
      description="from japan"
      id="green tea"
      key="green tea"
      note=""
    />
  </ol>
</div>
`;

exports[`render/component ExpenseList with no item 1`] = `
<div>
  <h2>
    Expense List
  </h2>
  <p>
    Tételek száma: 
    0
  </p>
  <h4>
    Nincs semmi kiadásod. Örüljél!
  </h4>
</div>
`;
